fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -786,6 +810,50 @@ protected  <T> CompletableFuture<Void> iterateRangeOnly(@Nonnull FDBRecordStore
                 });
     }
 
+    private CompletableFuture<Void> validateTypeStamp(@Nonnull FDBRecordStore store) {
+        final long minimalInterval = policy.getCheckIndexingMethodFrequencyMilliseconds();
+        if (minimalInterval < 0 || isScrubber) {
+            return AsyncUtil.DONE;
+        }
+        if (minimalInterval > 0) {
+            final long now = System.currentTimeMillis();
+            if (now < lastTypeStampCheckMillis + minimalInterval) {
+                return AsyncUtil.DONE;
+            }
+            lastTypeStampCheckMillis = now;
+        }
+        final IndexBuildProto.IndexBuildIndexingStamp expectedTypeStamp = getIndexingTypeStamp(store);
+        return forEachTargetIndex(index ->
+                store.loadIndexingTypeStampAsync(index)
+                        .thenAccept(typeStamp -> validateTypeStamp(typeStamp, expectedTypeStamp, index)));
+    }
+
+
+    private void validateTypeStamp(final IndexBuildProto.IndexBuildIndexingStamp typeStamp,
+                                   final IndexBuildProto.IndexBuildIndexingStamp expectedTypeStamp,
+                                   Index index) {
+        if (typeStamp == null && expectedTypeStamp.getMethod() == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+            // special case - null type stamp is considered a BY_RECORD
+            return;
+        }
+        if (typeStamp == null || typeStamp.getMethod() != expectedTypeStamp.getMethod() || isTypeStampBlocked(typeStamp)) {
+            throw new PartlyBuiltException(typeStamp, expectedTypeStamp,
+                    index.getName(), "Indexing stamp had changed",
+                    LogMessageKeys.INDEX_NAME, index,
+                    LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
+                    LogMessageKeys.INDEXER_ID, common.getUuid(),
+                    LogMessageKeys.EXPECTED, expectedTypeStamp,
```

Comment:
I'm not sure - we can expose the timestamps with `getSavedStamp` and `getExpectedStamp`, but exposing each of their fields will be hairy and would require constant maintenance for every future Stamp changes (+ the need to translate the `method` to another enum to avoid returning a protobuf value)
How about: 
```
ex.isBlocked()
ex.getSavedStamp()
ex.getExpectedStamp() 
```
(We currently only have the second one)

Reply:
Added a custom toString.